src/EnergyPlus/Pipes.hh

```diff
@@ -1,129 +1,53 @@
 		int BranchIndex; // Index of plant Branch index where this pipe resides
 		int CompIndex; // Index of plant Comp index where this pipe resides
-		bool OneTimeInit;
-		bool CheckEquipName;
-		bool EnvrnFlag;
 
 		// Default Constructor
 		LocalPipeData() :
-			TypeOf( 0 ),
 			InletNodeNum( 0 ),
 			OutletNodeNum( 0 ),
 			LoopNum( 0 ),
 			LoopSide( 0 ),
 			BranchIndex( 0 ),
-			CompIndex( 0 ),
-			OneTimeInit( true ),
-			CheckEquipName( true ),
-			EnvrnFlag( true )
+			CompIndex( 0 )
 		{}
 
-		// Member Constructor
-		LocalPipeData(
-			std::string const & Name, // main plant (cooling) loop ID
-			int const TypeOf, // type of pipe
-			int const InletNodeNum, // Node number on the inlet side of the plant
-			int const OutletNodeNum, // Node number on the inlet side of the plant
-			int const LoopNum, // Index of plant loop where this pipe resides
-			int const LoopSide, // Index of plant loop side where this pipe resides
-			int const BranchIndex, // Index of plant Branch index where this pipe resides
-			int const CompIndex, // Index of plant Comp index where this pipe resides
-			bool const OneTimeInit,
-			bool const CheckEquipName,
-			bool const EnvrnFlag
-		) :
-			Name( Name ),
-			TypeOf( TypeOf ),
-			InletNodeNum( InletNodeNum ),
-			OutletNodeNum( OutletNodeNum ),
-			LoopNum( LoopNum ),
-			LoopSide( LoopSide ),
-			BranchIndex( BranchIndex ),
-			CompIndex( CompIndex ),
-			OneTimeInit( OneTimeInit ),
-			CheckEquipName( CheckEquipName ),
-			EnvrnFlag( EnvrnFlag )
-		{}
+		static std::shared_ptr<PlantComponent> pipeFactory( std::string objectName );
+		int performEveryTimeInit();
+		int performOneTimeInit();
+		int performBeginEnvrnInit();
+		int simulate();
 
 	};
 
 	// Object Data
-	extern Array1D< LocalPipeData > LocalPipe; // dimension to number of pipes
-
-	// Functions
-
-	void
-	SimPipes(
-		int const CompType,
-		std::string & PipeName,
-		int & CompIndex,
-		Real64 const MaxVolFlowRate,
-		bool const InitLoopEquip,
-		bool const FirstHVACIteration
-	);
-
-	// End Plant Loop Module Driver Subroutines
-	//******************************************************************************
-
-	// Beginning of Plant Loop Module Get Input subroutines
-	//******************************************************************************
-
-	void
-	GetPipeInput();
-
-	// End of Get Input subroutines for the Plant Loop Module
-	//******************************************************************************
-
-	// Beginning Initialization Section of the Plant Loop Module
-	//******************************************************************************
-
-	void
-	InitializePipes(
-		int const PipeType, // Type of Pipe
-		std::string const & PipeName, // Name of Pipe
-		int & PipeNum, // Index into pipe structure for name
-		Real64 const MaxVolFlowRate // unused at present time
-	);
+	extern Array1D< std::shared_ptr< LocalPipeData > > LocalPipe; // dimension to number of pipes
```

Comment:
@jasondegraw Wrt indexing, agree on prefering iteration and range-based for loops. But we have a lot of index-based code and (non-container) multidimensional arrays need it. When you are doing array computations indexing is more natural and efficient (you don't want to be iterating over multiple iterators in parallel). For containers the use of indexing is less but there are times when you need to move through multiple containers in parallel so indexing is more efficient there too. I.e., we still need indexing. If we want to aim to be all modern and stuff how about a guideline that all-new code use 0-based container indexing and updates to existing code can use 1-based (when available). Having both is one advantage of wrapping `std::vector` in a collection.

There are a lot of other reasons for a collection system (or else [Boost Pointer Containers](http://www.boost.org/doc/libs/1_58_0/libs/ptr_container/doc/ptr_container.html) wouldn't exist). Not for local algorithm containers necessarily, but for the major model components they are a win for our developer pool (who are not as comfortable with C++ as you are):
* Containers of values and containers of pointers work completely differently: having a common API for them helps our developers.
* Is the pointer syntax of polymorphic containers a leaked artifact of their implementation? (Yes: anti-OO: ergo [Boost indirect_iterator](http://www.boost.org/doc/libs/1_58_0/libs/iterator/doc/indirect_iterator.html))
* What happens when you run std::sort on the vector of shared_ptr?!?
* Does `std::vector` guarantee to bounds check lookups in debug builds? (No.)
* Can developers figure out how to get the windows subset of surfaces and even for those that can isn't something we should package up with templatety goodness? (Yes, this could be a function if we don't have collection templates but it is nice to have a sub-collection hold its membership predicate for validation, refresh, ...)

I do want to get feedback on the collections framework from the team but it should consider the average, as well as expert, developer. I can share the design doc and code/examples with anyone interested.

[Another discussion, but `uniqe_ptr`, not `shared_ptr`, should be the default way to hold polymorphic types: only pay for shared ownership if you really need it.]